pkg/common-controller/snapshot_controller.go

```diff
@@ -171,53 +158,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
 func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnapshot) error {
```

Comment:
I feel the logic inside this function is kind of hard to follow. Trying to see whether the following top level functions (we can work out details of functions later) can help.

```
func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnapshot) error {
	klog.V(5).Infof("synchronizing VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))

	if snapshot.ObjectMeta.DeletionTimestamp != nil {
		return ctrl.syncDeletingSnapshot(snapshot)

	} else {
		klog.V(5).Infof("syncSnapshot[%s]: check if we should add finalizers on snapshot", utils.SnapshotKey(snapshot))
		ctrl.checkandAddSnapshotFinalizers(snapshot)

		if snapshot.Status == nil {
			return ctrl.syncUnBoundSnapshot(snapshot)
		} else {
			return ctrl.syncBoundSnapshot(snapshot)
		}
	}
	return nil
}

func (ctrl *csiSnapshotCommonController) syncUnBoundSnapshot(snapshot *crdv1.VolumeSnapshot) error {
	content, err := ctrl.getContentFromStore(snapshot)
	if err != nil {
		return err
	}
	
	if content != nil {
		addBoundFinalizer(snapshot)
		addSnapshotAsSourceFinalizer(snapshot)
		return updateSnapshotFromContent(snapshot, content)
	} else {
		if isStaticBinding(snapshot) {
			return nil
		}
		addBoundFinalizer(snapshot)
		return createSnapshotContent(snapshot)
	}

}

func (ctrl *csiSnapshotCommonController) syncBoundSnapshot(snapshot *crdv1.VolumeSnapshot) error {
	content, err := ctrl.getContentFromStore(snapshot)
	if err != nil {
		return err
	}
	
	if content == nil {
		return updateSnapshotWithNoContentError(snapshot)
	} else {
		return updateSnapshotFromContent(snapshot, content)
	}
}

```

Reply:
I initially thought about splitting syncSnapshot by whether snapshot.Status is nil or not, but I found that it will make the logic more complicated to follow.  If snapshot.Status != nil but it is empty or it is not empty but only with boundVolumeSnapshotContentName set but readyToUse is nil or false, and assume content != nil, we'll go ahead to update snapshot status from content status.  However content status could also be nil.  So we'll have to return without any update on snapshot status and wait for the next sync.  Eventually we'll get there but it will take longer.  It will also have duplicate logic to update status spreading in both syncSnapshotWithStatus and syncSnapshotWithoutStatus.

So I think the current logic is more straight-forward.

Reply:
Let's open an issue to resolve this, I  think the logic could be simplified, may deserve an issue to track separately.